Libraries/Alert/Alert.js

```diff
@@ -156,10 +169,53 @@ class Alert {
       // [TODO(macOS ISS#2323203)
     } else if (Platform.OS === 'macos') {
       const defaultInputs = [{default: defaultValue}];
-      AlertMacOS.prompt(title, message, callbackOrButtons, type, defaultInputs);
+      promptMacOS(title, message, callbackOrButtons, type, defaultInputs);
```

Comment:
Didn't add `modal` and `critical` arguments here, as the original implementation didn't have that. However, I can see that it may be necessary to add in the future, which would change the prompt function signature. Ideas?

Reply:
@HeyImChris @tom-un @NickGerleman How have you historically dealt with such deviating decision making?

Reply:
Long-term I think it would be best to refactor the `prompt` in RN core to have the same API as `alert`. This way we could fit all extra platform-specific options into the last `options` argument. It be a breaking change, but I may be worth it. Should also be pretty simple to write a codemod, although I have no experience in that yet (would be happy to contribute). If that makes sense to you, I can tag Eli from the RN core and see what's his thoughts

Reply:
There are a whole bunch of different solutions to public API deviation. The north star is to unify core APIs to be more desktop friendly. This is definitely the hardest option from both technical and non technical perspectives though. I've historically pushed pretty aggressively for doing this, as it saves a lot of pain in the long run, can benefit a greater ecosystem, and tends to produce higher quality changes.

For smaller components/Libares, something like AlertMacOS is not a particularly bad option. The benefit of having a separate component is that you can provide custom typings easily, and are insulated from upstream refactorings that tend to break forking in high traffic areas. Downside is you don't get upstream refactorings as well, and consumption is annoying. 

The really problematic cases I've seen have been for adding extra parameters to public methods, or extra methods/props. The story for discoverability and documentation is bad, and needing usage to be untyped has been really problematic for us.

I think this is more problematic for a team like Office though, where we simultaneously want to upgrade many projects across many repos to a new version of React Native at once. That sort of thing becomes really risky without good typings.

Reply:
Thanks @NickGerleman. This basically sums up my struggle with this issue, involving API decisions and balancing the ease of keeping the fork up to date.

I could throw a proposal on D&P repo to change the `prompt` API to be somewhat more generic:
```diff
-static prompt(title, message?, callbackOrButtons?, type?, defaultValue?, keyboardType?)
+static prompt(title, message?, callbackOrButtons?, options?)
```
which would allow to put android/ios/macos/windows specific options there. As you mentioned, this approach leads to less type safe code, since one could omit props that's specific for certain platform or pass the ones for another platform, thinking it works for theirs as well. Naming conventions like prefixing with `macos_` would help a bit.

Otherwise the sane decision would be to stay with `AlertMacOS` (or Alert.macos.js with an API of AlertMacOS), which provides the best type safety, but diverges from upstream implementation. However, since the issue was about deprecating AlertMacOS and going with Alert, I tried to figure out what's possible without it.

Let me know how and if we should proceed further :)

Reply:
This is a really good question and something that I've struggled with before.  Maybe I'm in the minority here but I find that forcing platforms to conform to an API set that doesn't make sense for the platform to not be worth the benefits of having one API.

Having "options" at the end is a good solution if we needed to keep the API identical, but I'm not convinced we need to do that. We can still #ifdef and share code between iOS/macOS when possible and also have reasonable APIs that make sense for the platforms they're on.

I always have to remind myself that RN isn't "write once, run anywhere", it's "learn once, write anywhere" so it's okay for platforms APIs to sometimes vary when it makes sense.